### ISSUE TITLE
Buffer overflow in indi_to_ped_fix() causes abort for long names

### ISSUE THREAD

**Issue by eggertj** [MarcNo/lifelines#323]
LifeLines 3.0.62 can abort when browsing an individual in ancestor view (!a) if an ancestor has a very long name.

It looks like the abort is happening in the function indi_to_ped_fix() in the file liflines/show.c.  It is trying to execute the line
    strcat(scratch, tmp1);
with the values of scratch being a string of length 90 and tmp1 being a string of length 20.  The value of len is 110 for this call.

This makes a string that is longer than 100 characters, the allocated length for scratch (when I ran the program), and causes an abort.  This appears to be a buffer overflow bug.  The problem is that the call to indi_to_ped_fix() specifies a length of 110, but the function only allows a length of 100.

I haven't checked later versions.

**Comment by memmerto** [comment]
This bug is present in the latest code.

**Comment by memmerto** [comment]
A fix is present in branch323 with a test (in llscr format).
Since we don't have a way (currently) to validate UI output this test has to be run manually.

**Comment by memmerto** [comment]
Fixed via https://github.com/MarcNo/lifelines/commit/afe246881eb070a9fa6fb41b2921a05656c14f76

**Comment by eggertj** [comment]
The fix didn’t work for me.  Changing one line makes it work, though:

In src/liflines/show.c, change line 699 from
```
        strncat(scratch, tmp1, ARRSIZE(scratch));
```
to
```
        strncat(scratch, tmp1, ARRSIZE(scratch) - strlen(name) - 1);
```

There might be a better way to do it, but that avoids the abort for me.

**Comment by memmerto** [comment]
You are right.  I was mis-remebering how the 'num' argument works for strncat.
(For whatever reason, this incorrect fix avoided the trap for me.)

I will fix this shortly.

**Comment by MichaelHendry** [comment]
It was my database which revealed this bug.
I was going to try to recompile with Jim Eggert's suggested fix (above), but I find that the copy of "src/liflines/show.c" in the 3.0.62 code I downloaded some years ago doesn't have the same line 699, indeed it has no "strncat".
My computer is an iMac, and I have a suspicion that the working version of 3.0.62 that I am using was supplied as an executable by a kind LINES-L forum member, as an attempt to configure and make the code on my computer results in a bewildering display (to me) of warnings and errors. Certainly, the working copy is dated Sept 2015 and the Makefiles are all dated Apr 2015.
I'm happy to help with testing, but would need some hand-holding.

**Comment by memmerto** [comment]
Fixed properly in 0f8d701d80ed4435103474c43bc400738aaf976b

**Comment by bgloyer** [comment]
It looks like this can still overflow if len is bigger than scratch which will happen if the terminal is wide.
       
```
	name = indi_to_name(indi, len - strlen(tmp1));
	strncpy(scratch, name, ARRSIZE(scratch));
	strncat(scratch, tmp1, ARRSIZE(scratch) - strlen(name) - 1);
```
I think I fixed it in  #330
